synapse/lib/stormlib/imap.py

```diff
@@ -0,0 +1,376 @@
+import asyncio
+
+import aioimaplib
+
+import synapse.exc as s_exc
+import synapse.lib.stormtypes as s_stormtypes
+
+async def run_imap_coro(coro):
+    '''
+    Raises or returns data
+    '''
+    try:
+        status, data = await coro
+    except asyncio.TimeoutError:
+        raise s_exc.TimeOut(mesg='Timed out waiting for IMAP server response.') from None
+
+    if status == 'OK':
+        return data
+
+    try:
+        mesg = data[0].decode()
+    except (TypeError, AttributeError, IndexError):
+        mesg = 'IMAP server returned an error'
+
+    raise s_exc.SynErr(mesg=mesg, status=status)
```

Comment:
Do we want to make a more specific exception here?

Reply:
Do you mean a custom class, e.g. IMAPErr?  The try block above tries to get the error message out of the response which is specific to the command.

Reply:
updated